File: LocalSearchOptimization/Examples/Problems/RectangularPacking/FloorplanSolution.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using LocalSearchOptimization.Components;
using LocalSearchOptimization.Examples.Structures.Permutation;
using LocalSearchOptimization.Examples.Structures.Tree;

namespace LocalSearchOptimization.Examples.RectangularPacking
{
    public class FloorplanSolution : IOrientedTree
    {
        private FloorplanProblem floorplanProblem;

        public int Transcoder { get; }

        public double CostValue { get; private set; }
        public int IterationNumber { get; set; }
        public double TimeInSeconds { get; set; }
        public bool IsCurrentBest { get; set; }
        public bool IsFinal { get; set; }

        public List<int> Order { get; }

        public List<bool> Branching { get; }

        public string InstanceTag { get; set; }

        public string OperatorTag { get; set; }

        public double[] X { get; }
        public double[] Y { get; }
        public double[] W { get; }
        public double[] H { get; }

        public double MaxWidth { get; set; }
        public double MaxHeight { get; set; }

        public int NumberOfRectangles { get => floorplanProblem.Dimension; }

        public FloorplanSolution(FloorplanProblem floorplanProblem) : this(floorplanProblem, Enumerable.Range(1, floorplanProblem.Dimension).ToList(), Enumerable.Repeat(false, floorplanProblem.Dimension).ToList().Concat(Enumerable.Repeat(true, floorplanProblem.Dimension)).ToList(), "init")
        {

        }

        public FloorplanSolution(FloorplanProblem floorplanProblem, List<int> order, List<bool> branching, string operatorName, int transcoder = 0)
        {
            this.floorplanProblem = floorplanProblem;
            Transcoder = transcoder;
            X = new double[floorplanProblem.Dimension + 1];
            Y = new double[floorplanProblem.Dimension + 1];
            W = transcoder % 2 == 0 ? floorplanProblem.W : floorplanProblem.H;
            H = transcoder % 2 == 0 ? floorplanProblem.H : floorplanProblem.W;
            Order = order;
            Branching = branching;
            OperatorTag = operatorName;
            DecodeSolution();
        }

        public IPermutation FetchPermutation(List<int> order, string operatorName)
        {
            return new FloorplanSolution(this.floorplanProblem, order, Branching, operatorName, Transcoder);
        }

        public ITreeBranching FetchBranching(List<bool> branching, string operatorName)
        {
            return new FloorplanSolution(this.floorplanProblem, Order, branching, operatorName, Transcoder);
        }

        public IOrientedTree FetchOrientedTree(List<int> order, List<bool> branching, string operatorName)
        {
            return new FloorplanSolution(this.floorplanProblem, order, branching, operatorName, Transcoder);
        }

        public ISolution Shuffle(int seed)
        {
            Random random = new Random(seed);
            List<bool> branching = new List<bool>();
            int opened = 0;
            int completed = 0;
            for (int i = 0; i < 2 * this.floorplanProblem.Dimension; i++)
            {
                if (completed < this.floorplanProblem.Dimension && (opened == 0 || random.NextDouble() < 0.5))
                {
                    branching.Add(false);
                    opened++;
                    completed++;
                }
                else
                {
                    branching.Add(true);
                    opened--;
                }
            }
            return new FloorplanSolution(this.floorplanProblem, Order.OrderBy(x => random.Next()).ToList(), branching, "shuffle", Transcoder);
        }

        public ISolution Transcode()
        {
            Stack<int> parents = new Stack<int>();
            parents.Push(0);
            List<int> uncoded = new List<int>(Order);
            List<int> order = new List<int>();
            List<bool> branching = new List<bool>();
            int opened = 0;
            while (uncoded.Count > 0)
            {
                int parent = parents.Peek();
                int next = 0;
                double nextX = int.MaxValue;
                foreach (int item in uncoded)
                {
                    if ((Y[item] == Y[parent] + H[parent])
                        && (X[parent] < X[item] + W[item])
                        && (X[item] < X[parent] + (parent > 0 ? W[parent] : MaxWidth))
                        && (X[item] < nextX))
                    {
                        next = item;
                    }
                }
                if (next > 0)
                {
                    order.Add(next);
                    branching.Add(false);
                    parents.Push(next);
                    uncoded.Remove(next);
                    opened++;
                }
                else
                {
                    branching.Add(true);
                    parents.Pop();
                    opened--;
                }
            }
            for (int i = 0; i < opened; i++)
            {
                branching.Add(true);
            }
            return new FloorplanSolution(this.floorplanProblem, order, branching, "transcode", (Transcoder + 1) % 4);
        }

        private void DecodeSolution()
        {
            LinearDecoder();
        }

        /// <summary>
        /// Decodes the tree structure into a packing layout. Trick with the contour structure provides O(n) performance that is the most efficient.
        /// </summary>
        private void LinearDecoder()
        {
            MaxWidth = 0;
            MaxHeight = 0;
            X[0] = 0;
            Y[0] = 0;
            W[0] = 0;
            H[0] = 0;
            LinkedList<int> countour = new LinkedList<int>();
            countour.AddLast(0);
            LinkedListNode<int> currentContour = countour.First;
            int perm = 0;
            foreach (bool b in Branching)
            {
                if (!b)
                {
                    int parent = currentContour.Value;
                    int current = Order[perm];
                    X[current] = X[parent] + W[parent];
                    LinkedListNode<int> currentTop = currentContour;
                    double maxY = 0;
                    bool topMostFound = false;
                    double rightSide = X[current] + W[current];
                    while (currentTop.Next != null && !topMostFound)
                    {
                        currentTop = currentTop.Next;
                        int top = currentTop.Value;
                        if (maxY < Y[top] + H[top]) maxY = Y[top] + H[top];

                        if (X[top] + W[top] <= rightSide)
                        {
                            currentTop = currentTop.Previous;
                            countour.Remove(top);
                        }
                        else topMostFound = true;
                    }
                    Y[current] = maxY;
                    countour.AddAfter(currentContour, current);
                    currentContour = currentContour.Next;
                    perm++;
                    double currentW = X[current] + W[current];
                    double currentH = Y[current] + H[current];
                    if (MaxWidth < currentW)
                        MaxWidth = currentW;
                    if (MaxHeight < currentH)
                        MaxHeight = currentH;
                }
                else
                {
                    currentContour = currentContour.Previous;
                }
            }
            //CostValue = Math.Pow(MaxWidth + 1.05 * MaxHeight, 2);
            CostValue = MaxWidth * MaxHeight;
        }

        /// <summary>
        /// Decodes the tree structure into a packing layout. Straightforward implementation provides O(n^2) performance. Used for demonstration purposes only. 
        /// </summary>
        private void QuadraticDecoder()
        {
            MaxWidth = 0;
            MaxHeight = 0;
            X[0] = 0;
            Y[0] = 0;
            Stack<int> parents = new Stack<int>();
            parents.Push(0);
            int perm = 0;
            foreach (bool b in Branching)
            {
                if (!b)
                {
                    int parent = parents.Peek();
                    int current = Order[perm];
                    X[current] = X[parent] + W[parent];
                    double maxY = 0;
                    for (int i = 0; i < perm; i++)
                    {
                        int p = Order[i];
                        if ((X[p] < X[current] + W[current]) && (X[current] < X[p] + W[p]) && (maxY < Y[p] + H[p]))
                        {
                            maxY = Y[p] + H[p];
                        }
                    }
                    Y[current] = maxY;
                    perm++;
                    parents.Push(current);
                    double currentW = X[current] + W[current];
                    double currentH = Y[current] + H[current];
                    if (MaxWidth < currentW)
                    {
                        MaxWidth = currentW;
                    }
                    if (MaxHeight < currentH)
                    {
                        MaxHeight = currentH;
                    }
                }
                else
                {
                    parents.Pop();
                }
            }
            CostValue = MaxWidth * MaxHeight;
        }

        public override string ToString()
        {
            StringBuilder s = new StringBuilder();
            StringBuilder b = new StringBuilder();
            StringBuilder t = new StringBuilder();
            foreach (int i in Order)
            {
                s.Append(i + ",");
            }

            int node = 0;
            foreach (bool i in Branching)
            {
                b.Append(i ? "true," : "false,");
                t.Append(i ? ")" : "(" + Order[node]);
                if (!i) node++;
            }
            s.Append("\n").Append(b);
            return s.ToString();
        }
    }
}

File: java/src/localsearchoptimization/examples/problems/rectangularpacking/FloorplanSolution.java
package localsearchoptimization.examples.problems.rectangularpacking;

import localsearchoptimization.components.ImageStyle;
import localsearchoptimization.components.Solution;
import localsearchoptimization.examples.structures.permutation.Permutation;
import localsearchoptimization.examples.structures.tree.OrientedTree;
import localsearchoptimization.examples.structures.tree.TreeBranching;

import java.awt.*;
import java.awt.image.BufferedImage;
import java.util.*;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class FloorplanSolution implements OrientedTree {

    private FloorplanProblem floorplanProblem;
    private int transcoder;
    private double cost;
    private int iterationNumber;
    private double timeInSeconds;
    private boolean isCurrentBest;
    private boolean isFinal;
    private int[] order;
    private boolean[] branching;
    private String instanceTag;
    private String operatorTag;

    public double[] x;
    public double[] y;
    public double[] w;
    public double[] h;

    public double maxWidth;
    public double maxHeight;

    public FloorplanSolution(FloorplanProblem floorplanProblem) {
        this(floorplanProblem, null, null, "init");
    }

    public FloorplanSolution(FloorplanProblem floorplanProblem, int[] order, boolean[] branching, String operatorName) {
        this(floorplanProblem, order, branching, operatorName, 0);
    }

    public FloorplanSolution(FloorplanProblem floorplanProblem, int[] order, boolean[] branching, String operatorName, int transcoder) {
        this.floorplanProblem = floorplanProblem;
        this.transcoder = transcoder;
        x = new double[floorplanProblem.dimension + 1];
        y = new double[floorplanProblem.dimension + 1];
        w = transcoder % 2 == 0 ? floorplanProblem.w : floorplanProblem.h;
        h = transcoder % 2 == 0 ? floorplanProblem.h : floorplanProblem.w;
        if (order != null)
            this.order = order;
        else
            this.order = IntStream.rangeClosed(1, floorplanProblem.dimension).toArray();
        if (branching != null)
            this.branching = branching;
        else {
            this.branching = new boolean[2 * floorplanProblem.dimension];
            boolean[] open = new boolean[floorplanProblem.dimension];
            boolean[] close = new boolean[floorplanProblem.dimension];
            Arrays.fill(open, false);
            Arrays.fill(close, true);
            System.arraycopy(open, 0, this.branching, 0, floorplanProblem.dimension);
            System.arraycopy(close, 0, this.branching, floorplanProblem.dimension, floorplanProblem.dimension);
        }
        this.operatorTag = operatorName;
        DecodeSolution();
    }

    public int numberOfRectangles() {
        return floorplanProblem.dimension;
    }

    public double itemsArea() {
        return floorplanProblem.totalArea;
    }

    public double utilization() {
        return floorplanProblem.totalArea * 100 / (maxWidth * maxHeight);
    }

    @Override
    public boolean[] branching() {
        return branching;
    }

    @Override
    public double cost() {
        return cost;
    }

    @Override
    public TreeBranching fetchBranching(boolean[] branching, String operationName) {
        return new FloorplanSolution(floorplanProblem, order, branching, operationName, transcoder);
    }

    @Override
    public OrientedTree fetchOrientedTree(int[] order, boolean[] branching, String operationName) {
        return new FloorplanSolution(floorplanProblem, order, branching, operationName, transcoder);
    }

    @Override
    public Permutation fetchPermutation(int[] order, String operationName) {
        return new FloorplanSolution(floorplanProblem, order, branching, operationName, transcoder);
    }

    @Override
    public int iterationNumber() {
        return iterationNumber;
    }

    @Override
    public void iterationNumber(int iteration) {
        iterationNumber = iteration;
    }

    @Override
    public double elapsedTime() {
        return timeInSeconds;
    }

    @Override
    public void elapsedTime(double seconds) {
        timeInSeconds = seconds;
    }

    @Override
    public boolean isCurrentBest() {
        return isCurrentBest;
    }

    @Override
    public void isCurrentBest(boolean isCurrentBest) {
        this.isCurrentBest = isCurrentBest;
    }

    @Override
    public boolean isFinal() {
        return isFinal;
    }

    @Override
    public void isFinal(boolean isFinal) {
        this.isFinal = isFinal;
    }

    @Override
    public String instanceTag() {
        return instanceTag;
    }

    @Override
    public void instanceTag(String tag) {
        instanceTag = tag;
    }

    @Override
    public String operatorTag() {
        return operatorTag;
    }

    @Override
    public int[] order() {
        return order;
    }

    @Override
    public Solution shuffle(int seed) {
        Random random = new Random(seed);
        boolean[] branching = new boolean[this.branching.length];
        int opened = 0;
        int completed = 0;
        int currentBranching = 0;
        for (int i = 0; i < 2 * floorplanProblem.dimension; i++) {
            if (completed < floorplanProblem.dimension && (opened == 0 || random.nextDouble() < 0.5)) {
                branching[currentBranching++] = false;
                opened++;
                completed++;
            } else {
                branching[currentBranching++] = true;
                opened--;
            }
        }
        int[] shuffled = order.clone();
        Collections.shuffle(Arrays.asList(shuffled), new Random(seed));
        return new FloorplanSolution(floorplanProblem, shuffled, branching, "shuffle", transcoder);
    }

    @Override
    public Solution transcode() {
        Stack<Integer> parents = new Stack<Integer>();
        parents.push(0);
        List<Integer> uncoded = IntStream.rangeClosed(1, floorplanProblem.dimension).boxed().collect(Collectors.toList());
        int[] order = new int[this.order.length];
        boolean[] branching = new boolean[this.branching.length];
        int opened = 0;
        int currentOrder = 0;
        int currentBranching = 0;
        while (uncoded.size() > 0) {
            int parent = parents.peek();
            int next = 0;
            double nextX = 0;
            for (int item : uncoded) {
                if ((y[item] == y[parent] + h[parent])
                        && (x[parent] < x[item] + w[item])
                        && (x[item] < x[parent] + (parent > 0 ? w[parent] : maxWidth))
                        && (x[item] >= nextX)) {
                    next = item;
                    nextX = x[item];
                }
            }
            if (next > 0) {
                order[currentOrder++] = next;
                branching[currentBranching++] = false;
                parents.push(next);
                uncoded.remove((Integer) next);
                opened++;
            } else {
                branching[currentBranching++] = true;
                parents.pop();
                opened--;
            }
        }
        for (int i = 0; i < opened; i++)
            branching[currentBranching++] = true;
        return new FloorplanSolution(floorplanProblem, order, branching, "transcode", (transcoder + 1) % 4);
    }

    private void DecodeSolution() {
        LinearDecoder();
    }

    public void LinearDecoder() {
        maxWidth = 0;
        maxHeight = 0;
        x[0] = 0;
        y[0] = 0;
        w[0] = 0;
        h[0] = 0;
        LinkedList<Integer> contourList = new LinkedList<Integer>();
        contourList.add(0);
        ListIterator<Integer> contour = contourList.listIterator();
        int perm = 0;
        for (boolean b : branching) {
            if (!b) {
                int parent = contour.next();
                int current = order[perm];
                x[current] = x[parent] + w[parent];
                double maxY = 0;
                boolean topMostFound = false;
                double rightSide = x[current] + w[current];
                while (contour.hasNext() && !topMostFound) {
                    int top = contour.next();
                    if (maxY < y[top] + h[top])
                        maxY = y[top] + h[top];
                    if (x[top] + w[top] <= rightSide)
                        contour.remove();
                    else topMostFound = true;
                }
                y[current] = maxY;
                if (topMostFound) contour.previous();
                contour.add(current);
                contour.previous();
                perm++;
                double currentW = x[current] + w[current];
                double currentH = y[current] + h[current];
                if (maxWidth < currentW)
                    maxWidth = currentW;
                if (maxHeight < currentH)
                    maxHeight = currentH;
            } else {
                contour.previous();
            }
        }
//        cost = maxWidth * maxHeight;
        cost = Math.pow(maxWidth + maxHeight, 2);
    }

    public BufferedImage draw(ImageStyle style) {
        double maxWidth = transcoder % 2 == 0 ? this.maxWidth : this.maxHeight;
        double maxHeight = transcoder % 2 == 0 ? this.maxHeight : this.maxWidth;
        double maxSize = Math.max(maxWidth, maxHeight);
        double scaleX = (style.imageWidth - style.marginX - 8 * style.radius) / maxSize;
        double scaleY = (style.imageHeight - style.marginY - 4 * style.radius) / maxSize;
        double[] x = transcoder % 2 == 0 ? this.x : this.y;
        double[] y = transcoder % 2 == 0 ? this.y : this.x;
        double[] w = transcoder % 2 == 0 ? this.w : this.h;
        double[] h = transcoder % 2 == 0 ? this.h : this.w;
        BufferedImage bitmap = new BufferedImage(style.imageWidth, style.imageHeight, BufferedImage.TYPE_INT_RGB);
        Graphics2D g = bitmap.createGraphics();
        g.setPaint(style.backgroundColor);
        g.fillRect(0, 0, bitmap.getWidth(), bitmap.getHeight());
        g.setStroke(new BasicStroke(style.penWidth));

        g.setPaint(Color.LIGHT_GRAY);
        g.fillRect((int) (style.marginX + style.penWidth), (int) (bitmap.getHeight() - style.penWidth - maxHeight * scaleY), (int) (maxWidth * scaleX), (int) (maxHeight * scaleY));

        int smallFontSize = (int) (0.9 * style.fontSize);
        g.setFont(new Font(style.fontName, Font.PLAIN, smallFontSize));

        for (int n : order) {
            int xScaled = (int) (style.marginX + style.penWidth + ((transcoder == 1 || transcoder == 2 ? maxWidth - x[n] - w[n] : x[n]) * scaleX));
            int yScaled = (int) (bitmap.getHeight() - style.penWidth - ((transcoder == 2 || transcoder == 3 ? maxHeight - y[n] : y[n] + h[n]) * scaleY));
            int wScaled = (int) (w[n] * scaleX);
            int hScaled = (int) (h[n] * scaleY);
            g.setPaint(new GradientPaint(xScaled, yScaled, style.backgroundColor, xScaled + wScaled, yScaled + hScaled, style.fillColor));
            g.fillRect(xScaled, yScaled, wScaled, hScaled);
            g.setPaint(style.penColor);
            g.drawRect(xScaled, yScaled, wScaled, hScaled);
            g.drawString(String.valueOf(n), xScaled + (wScaled - smallFontSize) / 2, yScaled + (hScaled + smallFontSize) / 2);
        }

        g.setPaint(Color.BLACK);
        g.setFont(new Font(style.fontName, Font.PLAIN, style.fontSize));
        g.drawString(String.format("Packing %1$s: %2$.4f (utilization %3$.2f%%)%4$s", cost == maxWidth * maxHeight ? "area" : "cost", cost, utilization(), isCurrentBest ? " <<<" : ""), 0, style.fontSize);
        g.drawString(String.format("Iterations: %1$d", iterationNumber), 0, 2 * style.fontSize);
        g.drawString(String.format("Time: %1$.3fs", timeInSeconds), 0, 3 * style.fontSize);
        return bitmap;
    }

    @Override
    public String toString() {
        return printTree();
    }
}